File: src/api/leads.ts
import { apiFetch } from './client';

// --- Tipos alineados con backend/app/schemas/lead.py ---

export type LeadStatus = 'new' | 'contacted' | 'visit_scheduled' | 'discarded' | 'rented';

export interface CreateLeadPayload {
  property_id: string;
  name: string;
  phone: string;
  email?: string;
  move_in_date?: string;
  budget?: number;
  has_pets?: boolean;
  has_guarantor?: boolean;
  message?: string;
}

export interface ApiLead {
  id: string;
  property_id: string;
  name: string;
  phone: string;
  email: string | null;
  move_in_date: string | null;
  /** Pydantic serializa Decimal como string en JSON. */
  budget: string | number | null;
  has_pets: boolean | null;
  has_guarantor: boolean | null;
  message: string | null;
  status: LeadStatus;
  created_at: string;
}

// --- Endpoints ---

export async function createLead(payload: CreateLeadPayload): Promise<ApiLead> {
  return apiFetch<ApiLead>('/api/v1/leads', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
}

export async function listLeads(propertyId: string): Promise<ApiLead[]> {
  return apiFetch<ApiLead[]>(`/api/v1/properties/${propertyId}/leads`);
}

export async function updateLeadStatus(leadId: string, status: LeadStatus): Promise<ApiLead> {
  return apiFetch<ApiLead>(`/api/v1/leads/${leadId}/status`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ status })
  });
}

File: backend/app/api/routes/leads.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.schemas.lead import LeadCreate, LeadOut, LeadStatusUpdate
from app.services import lead_service

router = APIRouter(tags=["leads"])


@router.post("/leads", response_model=LeadOut, status_code=201)
async def create_lead(
    payload: LeadCreate,
    db: AsyncSession = Depends(get_db),
) -> LeadOut:
    lead = await lead_service.create_lead(db, payload)
    return LeadOut.model_validate(lead)


@router.get("/properties/{property_id}/leads", response_model=list[LeadOut])
async def list_leads(
    property_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> list[LeadOut]:
    leads = await lead_service.list_leads(db, property_id)
    return [LeadOut.model_validate(lead) for lead in leads]


@router.patch("/leads/{lead_id}/status", response_model=LeadOut)
async def update_lead_status(
    lead_id: uuid.UUID,
    payload: LeadStatusUpdate,
    db: AsyncSession = Depends(get_db),
) -> LeadOut:
    lead = await lead_service.update_lead_status(db, lead_id, payload.status)
    return LeadOut.model_validate(lead)
